### PR TITLE
chore: handle PayForBlobNamespaceID in `merge`

### PR DESCRIPTION
**Pull request by rootulp**
Closes https://github.com/celestiaorg/celestia-app/issues/1285

## Description
1. Handle PayForBlobNamespaceID in `merge`
1. Extract `merge` to a new file
1. Extract `merge` to a new file
1. Add a unit test that uses sample block data

### CODE REVIEW

File: pkg/shares/testdata/sample-block.json
Comment (evan-forbes) [inline pkg/shares/testdata/sample-block.json:1]: how was this testdata generated? how do we regenerate it? 

admittedly, I'm not the biggest fan of using test data because the tests that rely on this could pass indefinitely, even if something in the block changes. imo, we should at least have a function to regenerate it a la https://github.com/celestiaorg/celestia-node/blob/3a1353c0d122e64b991983cf889530e2c7747d10/state/integration_test.go#L120-L147 and even then we'd still want some CI or something to notify us that it changed :grimacing: 

if we want to avoid import cycles, then I would prefer that we have those tests in a test package, but this is just a preference and doesn't have to be done in this PR

File: pkg/shares/reconstruct.go
Comment (evan-forbes) [inline pkg/shares/reconstruct.go:13]: [not blocking]
I understand changing this name, but it might be a tad confusing since celestia-node using the term reconstruction for a very different thing. 

Also, it deviates from the existing splitting merging terminology. Not going to block on it though, so will defer to you on that.

**Review by evan-forbes — COMMENTED** [review]
mind pointing out any changes involved in the fix? its a bit difficult to distinguish since we're moving code around.

if we wanted to get a little fancy, we could also modify an existing "fuzzing" integration test that is already generating random block data, to also test reconstruction.

File: pkg/shares/reconstruct.go
Comment (rootulp) [inline pkg/shares/reconstruct.go:35]: [for reviewers] this is the fix

File: pkg/shares/reconstruct.go
Comment (rootulp) [reply pkg/shares/reconstruct.go:13]: > Also, it deviates from the existing splitting merging terminology.

That was my intention because splitting and merging sound like inverse operations when they aren't
- **Split** converts block data into shares
- **Parse** converts shares into txs, blobs, or share sequences
- **Merge** or **Reconstruct** converts an extended data square into block data 

I can revert the rename if reviewers don't have alternative proposals

File: pkg/shares/testdata/sample-block.json
Comment (rootulp) [reply pkg/shares/testdata/sample-block.json:1]: > how was this testdata generated? how do we regenerate it?

https://github.com/celestiaorg/celestia-node/blob/3a1353c0d122e64b991983cf889530e2c7747d10/state/integration_test.go#L120-L147

> how do we regenerate it?

I can add that to celestia-app

File: pkg/shares/reconstruct.go
Comment (rootulp) [reply pkg/shares/reconstruct.go]: Resolved by reverting the rename.

**Comment by rootulp** [comment]
> if we wanted to get a little fancy, we could also modify an existing "fuzzing" integration test that is already generating random block data, to also test reconstruction.

I'm concerned the existing "fuzzing" test that tests reconstruction doesn't actually generate pay for blob txs [here](https://github.com/rootulp/celestia-app/blob/a42b83dac1627f428201ca98cea5914205ca5bfe/pkg/shares/merge_test.go#L101)

**Comment by evan-forbes** [comment]
> I'm concerned the existing "fuzzing" test that tests reconstruction doesn't actually generate pay for blob txs [here](https://github.com/rootulp/celestia-app/blob/a42b83dac1627f428201ca98cea5914205ca5bfe/pkg/shares/merge_test.go#L101)I'm concerned the existing "fuzzing" test that tests reconstruction doesn't actually generate pay for blob txs [here](https://github.com/rootulp/celestia-app/blob/a42b83dac1627f428201ca98cea5914205ca5bfe/pkg/shares/merge_test.go#L101)


whoops, I should have pointed to the test I was thinking of, yeah I think we should test this using blocks that are using more realistic data, like with this test.

https://github.com/celestiaorg/celestia-app/blob/main/app/test/fuzz_abci_test.go

**Review by evan-forbes — APPROVED** [review]

**Comment by evan-forbes** [comment]
created a follow up for this PR https://github.com/celestiaorg/celestia-app/issues/1322, in which case I think we're clear to merge

**Comment by rootulp** [comment]
Thanks for creating the follow-up issue. Will merge this as-is because it contains a fix.

**Pull request merged**
